samples/nrf9160/modem_shell/src/sock/sock.c

```diff
@@ -0,0 +1,1320 @@
+				sock_info_clear(socket_info);
+			}
+		}
+	}
+	shell_print(shell_global, "%s exit", __func__);
+}
+
+K_THREAD_DEFINE(sock_receive_thread, SOCK_RECEIVE_STACK_SIZE,
+		sock_receive_handler, NULL, NULL, NULL,
+		SOCK_RECEIVE_PRIORITY, 0, 0);
+
+int sock_recv(int socket_id, bool receive_start, int data_length, bool blocking,
+	      enum sock_recv_print_format print_format)
+{
+	struct sock_info_t *socket_info = get_socket_info_by_id(socket_id);
+
+	if (socket_info == NULL) {
+		return -EINVAL;
+	}
+
+	if (print_format != SOCK_RECV_PRINT_FORMAT_NONE) {
+		switch (print_format) {
+		case SOCK_RECV_PRINT_FORMAT_STR:
+		case SOCK_RECV_PRINT_FORMAT_HEX:
+			shell_print(
+				shell_global,
+				"Receive print format changed for socket id=%d",
+				socket_info->id);
+			socket_info->recv_print_format = print_format;
+			break;
+		default:
+			shell_error(
+				shell_global,
+				"Receive data print format (%d) must be %d or %d",
+				print_format,
+				SOCK_RECV_PRINT_FORMAT_STR,
+				SOCK_RECV_PRINT_FORMAT_HEX);
+			return -EINVAL;
+		}
+	} else if (receive_start) {
+		shell_print(shell_global,
+			    "Receive data calculation start socket id=%d", socket_info->id);
+		/* Set any leftover blocking sockets to non-blocking */
+		sock_all_set_nonblocking();
+		socket_info->recv_start_throughput = true;
+		socket_info->recv_data_len = 0;
+		socket_info->log_receive_data = false;
+		socket_info->start_time_ms = 0;
+		socket_info->recv_end_time_ms = 0;
+		set_sock_blocking_mode(socket_info->fd, blocking);
+
+		/* Set expected data length when summary is printed */
+		if (data_length > 0) {
+			socket_info->recv_data_len_expected = data_length;
+		} else {
+			socket_info->recv_data_len_expected = SOCK_DATA_LEN_NONE;
+		}
+	} else {
+		print_throughput_summary(
+			socket_info->recv_data_len,
+			socket_info->recv_end_time_ms - socket_info->start_time_ms);
+	}
+	return 0;
+}
+
+int sock_close(int socket_id)
+{
+	struct sock_info_t *socket_info = get_socket_info_by_id(socket_id);
+
+	if (socket_info == NULL) {
+		return -EINVAL;
+	}
+	shell_print(shell_global, "Close socket id=%d, fd=%d", socket_info->id, socket_info->fd);
+	sock_info_clear(socket_info);
+	return 0;
+}
+
+int sock_rai_enable(int rai_enable)
+{
+	enum at_cmd_state state = AT_CMD_OK;
+	char command[] = "AT%%RAI=0";
+	int err;
+
+	if (rai_enable == SOCK_RAI_NONE) {
+		shell_error(shell_global, "No valid RAI options given");
+		return -EINVAL;
+	}
+	sprintf(command, "AT%%RAI=%d", rai_enable);
+	err = at_cmd_write(command, NULL, 0, &state);
+	if (state == AT_CMD_OK) {
+		shell_print(
+			shell_global,
+			"Release Assistance Indication functionality set to enabled=%d",
+			rai_enable);
+	} else {
+		shell_error(shell_global, "Error state=%d, error=%d",
+			    state, err);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int sock_get_nrf_fd_by_zephyr_fd(int zephyr_fd)
+{
+	/* Returned pointer is fd used in NRF side */
+	int nrf_fd = (int)z_get_fd_obj(zephyr_fd, NULL, 1);
```

Comment:
This won't work anymore, as the `obj` associated with the nrf socket is a structure now:
https://github.com/nrfconnect/sdk-nrf/blob/master/lib/nrf_modem_lib/nrf91_sockets.c#L43
So basically you should reach the `nrf_fd` field of the structure.

Reply:
Also note, that it should be no longer needed after https://github.com/nrfconnect/sdk-nrf/pull/4528/files is in - you'll be able to use RAI options with regular socket descriptor.

Reply:
This is good. So we still need to keep our way for some time.

Reply:
We need to wait for Zephyr upmerge and the PR 4528 and then do our changes. Zephyr side we need even for compilation.

Reply:
Zephyr changes are there so I've fixed this RAI support in MOSH.